app/javascript/shared/components/tubeScanValidators.js

```diff
@@ -53,16 +55,25 @@ const checkDuplicates = (tubeList) => {
       return { valid: false, message: 'Barcode has been scanned multiple times' }
     }
     else {
-      return validTubeScanMessage()
+      return validScanMessage()
     }
   }
 }
 
-// Renders a valid Tube scan
-// 
-// Returns an object with the right tube scan validation message
-const validTubeScanMessage = () => {
-  return { valid: true, message: 'Great!' }
+// Returns a validator that ensures the purpose namess of all the tubes match
+// the name for the one provided. Typically the one provided should be the one
+// of the purposes from the full set of tubes being validated.
+const checkMatchingPurposes = (purpose) => {
+  return (tube) => {
+    if (tube && purpose && tube.purpose.name !== purpose.name) {
```

Comment:
We could use the safe navigation operator here: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Also, with the current code if you *do* end up with a tube without a purpose (unfortunately they exist) the error message generation will currently fail. I've been using 'UNKNOWN' in the presenters in the case a tube has no purpose. (I'm hoping we won't see them)

Reply:
I've used the safe navigation on the `purpose` for the given `tube` but I don't think we should be removing the other checks for `tube` and `purpose` being defined.  The reason is covered in the unit tests.

The `purpose` that was passed in could be `undefined` and this is not an error case.  This happens when the user hasn't provided any tubes yet and then they provide their first tube.  No tube can be valid until at least one is entered and the purpose to match against is taken from valid tubes so you have a chicken and egg unless you allow all tubes to pass when the purpose is `undefined`. 

Similarly, the `tube` can be `undefined` and that's legitimate behaviour.  The empty tube rack positions do submit for validation and they have no tube yet.